### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by luizaandrade**
I think this is ready for publication!!! \o/

### CODE REVIEW

**Review by luisesanmartin — APPROVED** [review]
@luizaandrade I'm merging these changes to main. I also changed the readme file so it's in line with the options we now have

**Comment by luisesanmartin** [comment]
@kbjarkefur the code in this branch is ready for review

**Comment by luisesanmartin** [comment]
@bbdaniels this is ready for review. Note that all the issues with the tag `v1.0` should be solved in this branch

**Comment by bbdaniels** [comment]
- [ ] Runfile crash:
```
.   // Lint with results in excel file
.   lint "${test_dir}/bad.do", nosummary          ///
>     excel("${test_dir}/detect_lint.xlsx")

Linting file: /Users/bbdaniels/GitHub/stata-linter/test/bad.do


Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bbdaniels/Library/Application Support/Stata/ado/plus/py/stata_linter_detect.py", line 739, in stata_linter_detect_py
    with pd.ExcelWriter(excel) as writer:
  File "/usr/local/Caskroom/miniconda/base/lib/python3.8/site-packages/pandas/io/excel/_openpyxl.py", line 48, in __init__
    from openpyxl.workbook import Workbook
ModuleNotFoundError: No module named 'openpyxl'
r(7102);
```

**Comment by bbdaniels** [comment]
- [ ] Hard tabs do not appear to be replaced except when they are at the beginning of a line, leading to odd spacing after partial correctio
Screen Shot(3)
n

**Comment by bbdaniels** [comment]
- [ ] Confirming this is intentional: long lines are not broken when they fall inside functions. If correct, please add note to helpfile confirming exceptions

Screen Shot

**Comment by bbdaniels** [comment]
- [ ] Line lengths remain too long after correction, possibly a collision here with hard tab replacement plus the extra characters from `\\\`

Screen Shot(4)

**Comment by bbdaniels** [comment]
- [ ] Add informative error when invalid response given

Screen Shot(5)

**Comment by bbdaniels** [comment]
- [ ] Hard tabs not replaced within comments

Screen Shot(6)

**Comment by bbdaniels** [comment]
- [ ] `detect` function flags math symbols within `" "` when it should not do so; `correct` ignores them

**Comment by bbdaniels** [comment]
- [ ] Existing indentation is not respected in re-indentation function

Screen Shot(7)

**Comment by bbdaniels** [comment]
- [ ] `inprep` is an unusual option name that is not intuitive, why isn't this something like `force`?

**Comment by bbdaniels** [comment]
- [ ] The help file should separate clearly what issues are only caught in detection and which issues are actually corrected, and what corrections are and are not done

**Comment by bbdaniels** [comment]
- [ ] The help file should also note common confusions that the code cannot distinguish. For example, under "Check if backslashes are not used in file paths", the code is only looking for backslashes in general because it cannot tell what is a file path. Therefore these are "flags" that could be errors, but cannot be accurately detected by the code and cannot be corrected. More description for each of these items in this line would be useful.

**Review by bbdaniels — CHANGES_REQUESTED** [review]
Ok I have tested this as thoroughly as I can -- most of the functionality looks to be in very good shape. I caught a few edge cases in my runs as flagged above, and I think the help file could use a little more explanation of each of the tests and what the command does and does not do for each one. I am pushing some minor changes to the first part of the help file.

**Comment by luisesanmartin** [comment]
> * [ ]  Runfile crash, looks like a Python dependency issue:
> 
> 
> ```
> .   // Lint with results in excel file
> .   lint "${test_dir}/bad.do", nosummary          ///
> >     excel("${test_dir}/detect_lint.xlsx")
> 
> Linting file: /Users/bbdaniels/GitHub/stata-linter/test/bad.do
> 
> 
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/bbdaniels/Library/Application Support/Stata/ado/plus/py/stata_linter_detect.py", line 739, in stata_linter_detect_py
>     with pd.ExcelWriter(excel) as writer:
>   File "/usr/local/Caskroom/miniconda/base/lib/python3.8/site-packages/pandas/io/excel/_openpyxl.py", line 48, in __init__
>     from openpyxl.workbook import Workbook
> ModuleNotFoundError: No module named 'openpyxl'
> r(7102);
> ```

@bbdaniels when have time, can you pull from this branch and try again? it should a message saying that you need to install the Python package `openpyxl`

**Comment by luisesanmartin** [comment]
> * [ ]  Hard tabs do not appear to be replaced when they are not at the beginning of a line, leading to odd spacing after partial correction
> 
> 
> Screen Shot(3)

corrected in faa55e1

**Comment by luisesanmartin** [comment]
> * [x]  Confirming this is intentional: long lines are not broken when they fall inside functions. If correct, please add note to helpfile confirming exceptions
> 
> 
> Screen Shot

yes, this is intentional. Lines are not split in white spaces inside parentheses, curly brackets, or double quotes. Added a clarification on this in 2552456

**Comment by luisesanmartin** [comment]
> * [x]  Line lengths remain too long after correction, possibly a collision here with hard tab replacement plus the extra characters from `///`
> 
> 
> Screen Shot(4)

thanks, this was actually a bug that I hadn't detected. It was corrected in 7d238e0

**Comment by luisesanmartin** [comment]
> * [x]  Add informative error when invalid response given
> 
> 
> Screen Shot(5)

added in 97ca9dd
image

**Comment by luisesanmartin** [comment]
> * [x]  Hard tabs not replaced within comments
> 
> 
> Screen Shot(6)

corrected in 5aad6d9

**Comment by luisesanmartin** [comment]
> * [x]  `detect` function flags math symbols within `" "` when it should not do so; `correct` ignores them

corrected the function in 2de4ca8 so `detect` knows which portions of a line are not within quotes

**Comment by luisesanmartin** [comment]
> * [ ]  Existing indentation is not respected in re-indentation function
> 
> 
> Screen Shot(7)

Checking this was challenging but I could finally detect what was wrong. A few comments:
- specifically in this case, the command is checking which lines inside the loop don't have the correct loop indentation, which by default is 4. Then it's moving every line that doesn't have at least a 4-indentation inside the loop so they have an indentation of 4 + the indentation of the line that opens the loop. Since you're using an indentation of 2, your case would have worked better with the option `indentation(2)` which changes the default of 4
- That said, the correction of indentation only works with respect of the indentation of the line that opens the loop. In short, it will check which lines inside the loop don't have an indentation of at least (indentation of opening line) + (indentation level -- 4 by default) and will change their indentation to that number if theirs is lower. Correction of indentation doesn't work with respect of any extra indentation that should be ideally added because the previous line ended with a `///`, which I understand is what you were testing with this command. Bad indentations with respect of continued lines (`///`) are only flagged in the detection function, but they are not corrected. We might add that functionality in a later version though
- Reviewing the code for this comment made me discover a bug: the command was only checking the correct indentation of the first non-blank line of the loop. Now it checks all the lines until it's closed and works well with embedded loops (added in 6119628)

**Comment by luisesanmartin** [comment]
> * [x]  `inprep` is an unusual option name that is not intuitive, why isn't this something like `force`?

corrected in 5f83d6f

File: src/lint.ado [hunk 1]
```diff
@@ -260,77 +338,225 @@ program lint
...
+		* Stata result footer
+		if !missing("`footer'") {
+
+				display as result 	_dup(85) "-"
+
+			if "`excel'" != "" {
+				display as result 	`"{phang}File {browse "`excel'":`excel'} created.{p_end}"'
+			}
+
+				display as result 	`"{phang}For more information about coding guidelines visit the {browse "https://https://dimewiki.worldbank.org/Stata_Linter":Stata linter wiki.}{p_end}"'
```
Comment (kbjarkefur) [inline src/lint.ado:412]:
```suggestion
				display as result 	`"{phang}For more information about coding guidelines visit the {browse "https://dimewiki.worldbank.org/Stata_Linter":Stata linter wiki.}{p_end}"'
```

File: LICENSE [hunk 2]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2020 Mizuhiro Suzuki
+Copyright (c) 2022 Mizuhiro Suzuki, World Bank Group.
```
Comment (kbjarkefur) [inline LICENSE:3]:
```suggestion
Copyright (c) DIME Analytics, DIME, DEC, The World Bank Group.
```
I saw my name was listed on `ietoolkit`. That is not right. All these licenses should be DA, DIME, WB.
